### PR TITLE
Add a `kubernetes-log4j` module

### PR DESCRIPTION
**Pull request by ppkarwasz**
## Description

The `kubernetes-log4j` module adds the ability to Log4j Core to use Kubernetes attributes in a configuration file.

It is a cleaned-up version of the [`org.apache.logging.log4j:log4j-kubernetes`](https://logging.apache.org/log4j/2.x/log4j-kubernetes.html), which is also released under ASL 2.0.

As explained in #5682, it does make more sense to host is here since:

 * it only depends on a very stable `StrLookup` dependency from `log4j-core`,
 * the number and kind of properties available through `kubernetes-client` depend on its version.

## Type of change
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/main/CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [x] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/main/doc/CHEATSHEET.md) accordingly
 - [x] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Comment by ppkarwasz** [comment]
For compatibility with `org.apache.logging.log4j:log4j-kubernetes` the lookup allows to override the automatic Kubernetes Client configuration using Log4j property sources. If this does not make sense in this project, I can remove the `ClientBuilder` and `ClientProperties` classes.

**Comment by ppkarwasz** [comment]
I am converting this to draft, since I need to add tests for most of the code.

Do you have an util to replace the `ContainerUtil` from the original `log4j-kubernetes`? The class tries to determine the id of the Kubernetes container, by parsing `/proc/self/cgroup` on Linux, which is error prone and not easy to test.

File: log4j/pom.xml [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+          <execution>
+            <id>generate-osgi-bundle</id>
+            <goals>
+              <goal>bundle</goal>
+            </goals>
+          </execution>
+        </executions>
+      </plugin>
+    </plugins>
+  </build>
```
Comment (manusa) [inline log4j/pom.xml:88]: I'm not 100% sure if these should be required as they are defined in the parent and you're already defining some of the osgi properties.

**Comment by manusa** [comment]
>  If this does not make sense in this project, I can remove the ClientBuilder and ClientProperties classes.

I'm not sure how does this play with our own property inference logic:

https://github.com/fabric8io/kubernetes-client/blob/315f4869a0cfa658099270dfea0f4d143cc99899/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/Config.java#L409

I assume that for backwards-compatibility this is necessary, so I'm fine keeping it.

**Comment by manusa** [comment]
> Do you have an util to replace the ContainerUtil from the original log4j-kubernetes?

Not that I know of.

File: log4j/src/test/java/io/fabric8/kubernetes/log4j/lookup/KubernetesLookupTest.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    final ObjectMeta meta = new ObjectMeta();
+    final Map<String, String> annotations = new HashMap<>();
+    annotations.put("test", "name");
+    meta.setAnnotations(annotations);
+    final Map<String, String> labels = new HashMap<>();
+    labels.put("ns", "my-namespace");
+    meta.setLabels(labels);
+    meta.setUid(UUID.randomUUID().toString());
+    namespace.setMetadata(meta);
+    return namespace;
```
Comment (manusa) [inline log4j/src/test/java/io/fabric8/kubernetes/log4j/lookup/KubernetesLookupTest.java:90]: nit: You could use NamespaceBuilder fluent interface to make this a little bit more readable and avoid declaring intermediate variables (meta, annotations, and so on).

File: pom.xml [hunk 3]
```diff
@@ -221,6 +222,7 @@
     <module>java-generator</module>
     <module>httpclient-okhttp</module>
     <module>httpclient-vertx</module>
+    <module>log4j</module>
```
Comment (manusa) [inline pom.xml:225]: To be able to properly generate our BOM this module should also be added as a dependency to the dependencyManagement section.

File: log4j/pom.xml [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+  <dependencies>
+    <dependency>
+      <groupId>io.fabric8</groupId>
+      <artifactId>kubernetes-client</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>io.fabric8</groupId>
+      <artifactId>kubernetes-model-core</artifactId>
+    </dependency>
```
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply log4j/pom.xml]: I removed it.

File: log4j/pom.xml (hunk 1 above)
Comment (ppkarwasz) [reply log4j/pom.xml]: I replaced the artifact type from `jar` to `bundle`, in order to activate the `maven-bundle-plugin`.

File: log4j/src/test/java/io/fabric8/kubernetes/log4j/lookup/KubernetesLookupTest.java (hunk 2 above)
Comment (ppkarwasz) [reply log4j/src/test/java/io/fabric8/kubernetes/log4j/lookup/KubernetesLookupTest.java]: Nice trick, thanks

File: pom.xml [hunk 5]
```diff
@@ -700,6 +702,15 @@
         <type>pom</type>
       </dependency>
 
+      <!-- Log4j dependencies, imported as a BOM -->
+      <dependency>
+        <groupId>org.apache.logging.log4j</groupId>
+        <artifactId>log4j-bom</artifactId>
+        <version>${log4j.version}</version>
+        <scope>import</scope>
+        <type>pom</type>
```
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply pom.xml]: No, it is not necessary. I moved the only required dependency (`log4j-core`) to the module itself.

File: pom.xml (hunk 3 above)
Comment (ppkarwasz) [reply pom.xml:224]: Done

**Comment by ppkarwasz** [comment]
@manusa,

Thank you for your review. I finally had some time to work on this PR.

Apart from addressing your remarks above:

- I introduced a system property `kubernetes.log4j.useProperties`. When set to `false` (default) Log4j properties are not used to configure the Kubernetes client. I think that this is a good compromise to maintain compatibility with `log4j-kubernetes`, but require users to set a special property to enable it.
- I increased the test coverage.

**Comment by ppkarwasz** [comment]
The OpenShift failures seem unrelated to my PR.

**Comment by rohanKanojia** [comment]
@ppkarwasz : Is it possible to add some README or user documentation in doc folder regarding how to use this?

**Comment by ppkarwasz** [comment]
@rohanKanojia,

I have added a `doc/KubernetesLog4j.md` file in the `/doc` folder.

**Review by oscerd — APPROVED** [review]

**Review by manusa — APPROVED** [review]
LGTM, thx!

**Comment by manusa** [comment]
> @manusa,
> 
> Thank you for offering to maintain this component. If some Log4j-specific issues were to appear with it, feel free to `@`-mention me.

Will do :) Thanks for the contribution!

**Pull request merged**